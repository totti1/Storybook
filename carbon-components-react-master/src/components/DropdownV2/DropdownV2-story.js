import React from 'react';
import { storiesOf, action } from '@storybook/react';
import DropdownV2 from '../DropdownV2';

const items = [
  {
    id: 'option-1',
    text: 'Option 1',
  },
  {
    id: 'option-2',
    text: 'Option 2',
  },
  {
    id: 'option-3',
    text: 'Option 3',
  },
  {
    id: 'option-4',
    text: 'Option 4',
  },
];

storiesOf('DropdownV2', module)
  .addWithInfo(
    'default',
    `
    DropdownV2
  `,
    () => (
      <div style={{ width: 300 }}>
        <DropdownV2
          label="Label"
          items={items}
          itemToString={item => (item ? item.text : '')}
          onChange={action('onChange')}
        />
      </div>
    )
  )
  .addWithInfo(
    'inline',
    `
    Disabled DropdownV2
  `,
    () => (
      <div style={{ width: 300 }}>
        <DropdownV2
          type="inline"
          label="Label"
          items={items}
          itemToString={item => (item ? item.text : '')}
          onChange={action('onChange')}
        />
      </div>
    )
  )
  .addWithInfo(
    'disabled',
    `
    Disabled DropdownV2
  `,
    () => (
      <div style={{ width: 300 }}>
        <DropdownV2
          label="Label"
          items={items}
          itemToString={item => (item ? item.text : '')}
          onChange={action('onChange')}
          disabled
        />
      </div>
    )
  )
  .addWithInfo(
    'disabled - inline',
    `
    Disabled Inline DropdownV2
  `,
    () => (
      <div style={{ width: 300 }}>
        <DropdownV2
          type="inline"
          label="Label"
          items={items}
          itemToString={item => (item ? item.text : '')}
          onChange={action('onChange')}
          disabled
        />
      </div>
    )
  );
