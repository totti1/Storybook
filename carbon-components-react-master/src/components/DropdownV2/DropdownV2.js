import cx from 'classnames';
import Downshift from 'downshift';
import PropTypes from 'prop-types';
import React from 'react';
import ListBox, { PropTypes as ListBoxPropTypes } from '../ListBox';

export default class DropdownV2 extends React.Component {
  static propTypes = {
    /**
     * Disable the control
     */
    disabled: PropTypes.bool,

    /**
     * We try to stay as generic as possible here to allow individuals to pass
     * in a collection of whatever kind of data structure they prefer
     */
    items: PropTypes.array.isRequired,

    /**
     * Allow users to pass in arbitrary items from their collection that are
     * pre-selected
     */
    initialSelectedItem: PropTypes.object,

    /**
     * Helper function passed to downshift that allows the library to render a
     * given item to a string label. By default, it extracts the `label` field
     * from a given item to serve as the item label in the list.
     */
    itemToString: PropTypes.func,

    /**
     * `onChange` is a utility for this controlled component to communicate to a
     * consuming component what kind of internal state changes are occuring.
     */
    onChange: PropTypes.func,

    /**
     * Generic `label` that will be used as the textual representation of what
     * this field is for
     */
    label: PropTypes.node.isRequired,

    type: ListBoxPropTypes.ListBoxType,
  };

  static defaultProps = {
    disabled: false,
    type: 'default',
    itemToString: item => (item ? item.label : ''),
  };

  handleOnChange = selectedItem => {
    if (this.props.onChange) {
      this.props.onChange({ selectedItem });
    }
  };

  render() {
    const {
      className: containerClassName,
      disabled,
      items,
      label,
      itemToString,
      type,
      initialSelectedItem,
    } = this.props;
    const className = cx('bx--dropdown-v2', containerClassName);
    return (
      <Downshift
        onChange={this.handleOnChange}
        itemToString={itemToString}
        defaultSelectedItem={initialSelectedItem}>
        {({
          isOpen,
          itemToString,
          selectedItem,
          highlightedIndex,
          getRootProps,
          getButtonProps,
          getItemProps,
          getLabelProps,
        }) => (
          <ListBox
            type={type}
            className={className}
            disabled={disabled}
            {...getRootProps({ refKey: 'innerRef' })}>
            <ListBox.Field {...getButtonProps({ disabled })}>
              <span className="bx--list-box__label" {...getLabelProps()}>
                {selectedItem ? itemToString(selectedItem) : label}
              </span>
              <ListBox.MenuIcon isOpen={isOpen} />
            </ListBox.Field>
            {isOpen && (
              <ListBox.Menu>
                {items.map((item, index) => (
                  <ListBox.MenuItem
                    key={itemToString(item)}
                    isActive={selectedItem === item}
                    isHighlighted={highlightedIndex === index}
                    {...getItemProps({ item, index })}>
                    {itemToString(item)}
                  </ListBox.MenuItem>
                ))}
              </ListBox.Menu>
            )}
          </ListBox>
        )}
      </Downshift>
    );
  }
}
