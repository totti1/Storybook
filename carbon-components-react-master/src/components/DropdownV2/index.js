export default from './DropdownV2';
