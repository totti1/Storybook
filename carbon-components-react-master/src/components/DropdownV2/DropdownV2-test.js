import React from 'react';
import { mount } from 'enzyme';
import {
  assertMenuOpen,
  assertMenuClosed,
  findMenuItemNode,
  openMenu,
  generateItems,
  generateGenericItem,
} from '../ListBox/test-helpers';
import DropdownV2 from '../DropdownV2';

describe('DropdownV2', () => {
  let mockProps;
  beforeEach(() => {
    mockProps = {
      items: generateItems(5, generateGenericItem),
      onChange: jest.fn(),
      label: 'input',
      placeholder: 'Filter...',
      type: 'default',
    };
  });

  it('should render', () => {
    const wrapper = mount(<DropdownV2 {...mockProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should initially render with the menu not open', () => {
    const wrapper = mount(<DropdownV2 {...mockProps} />);
    assertMenuClosed(wrapper);
  });

  it('should let the user open the menu by clicking on the control', () => {
    const wrapper = mount(<DropdownV2 {...mockProps} />);
    openMenu(wrapper);
    assertMenuOpen(wrapper, mockProps);
  });

  it('should let the user select an option by clicking on the option node', () => {
    const wrapper = mount(<DropdownV2 {...mockProps} />);
    openMenu(wrapper);
    findMenuItemNode(wrapper, 0).simulate('click');
    expect(mockProps.onChange).toHaveBeenCalledTimes(1);
    expect(mockProps.onChange).toHaveBeenCalledWith({
      selectedItem: mockProps.items[0],
    });
    assertMenuClosed(wrapper);

    mockProps.onChange.mockClear();

    openMenu(wrapper);
    findMenuItemNode(wrapper, 1).simulate('click');
    expect(mockProps.onChange).toHaveBeenCalledTimes(1);
    expect(mockProps.onChange).toHaveBeenCalledWith({
      selectedItem: mockProps.items[1],
    });
  });
});
