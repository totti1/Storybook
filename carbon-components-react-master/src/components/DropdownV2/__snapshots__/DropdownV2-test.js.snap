// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownV2 should render 1`] = `
<DropdownV2
  disabled={false}
  itemToString={[Function]}
  items={
    Array [
      Object {
        "id": "id-0",
        "label": "Item 0",
        "value": 0,
      },
      Object {
        "id": "id-1",
        "label": "Item 1",
        "value": 1,
      },
      Object {
        "id": "id-2",
        "label": "Item 2",
        "value": 2,
      },
      Object {
        "id": "id-3",
        "label": "Item 3",
        "value": 3,
      },
      Object {
        "id": "id-4",
        "label": "Item 4",
        "value": 4,
      },
    ]
  }
  label="input"
  onChange={[MockFunction]}
  placeholder="Filter..."
  type="default"
>
  <Downshift
    breakingChanges={Object {}}
    defaultHighlightedIndex={null}
    defaultInputValue=""
    defaultIsOpen={false}
    defaultSelectedItem={null}
    environment={[Window]}
    getA11yStatusMessage={[Function]}
    id="downshift-1"
    itemToString={[Function]}
    onChange={[Function]}
    onInputValueChange={[Function]}
    onOuterClick={[Function]}
    onSelect={[Function]}
    onStateChange={[Function]}
    onUserAction={[Function]}
    selectedItemChanged={[Function]}
  >
    <ListBox
      className="bx--dropdown-v2"
      disabled={false}
      innerRef={[Function]}
      type="default"
    >
      <div
        className="bx--dropdown-v2 bx--list-box bx--form-item"
      >
        <ListBoxField
          aria-expanded={false}
          aria-haspopup={true}
          aria-label="open menu"
          disabled={false}
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
        >
          <div
            aria-expanded={false}
            aria-haspopup={true}
            aria-label="open menu"
            className="bx--list-box__field"
            disabled={false}
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex="0"
          >
            <span
              className="bx--list-box__label"
              htmlFor="downshift-input-2"
            >
              input
            </span>
            <ListBoxMenuIcon
              isOpen={false}
              translateWithId={[Function]}
            >
              <div
                className="bx--list-box__menu-icon"
              >
                <Icon
                  alt="Open menu"
                  description="Open menu"
                  fillRule="evenodd"
                  name="caret--down"
                  role="img"
                >
                  <svg
                    alt="Open menu"
                    aria-label="Open menu"
                    fillRule="evenodd"
                    height="5"
                    name="caret--down"
                    role="img"
                    viewBox="0 0 10 5"
                    width="10"
                  >
                    <title>
                      Open menu
                    </title>
                    <path
                      d="M10 0L5 5 0 0z"
                      key="key0"
                    />
                  </svg>
                </Icon>
              </div>
            </ListBoxMenuIcon>
          </div>
        </ListBoxField>
      </div>
    </ListBox>
  </Downshift>
</DropdownV2>
`;
